grpc-ballerina/auth_utils.bal

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2021 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/log;
+import ballerina/stringutils;
+
+# Represents the Authorization header name.
+public const string AUTH_HEADER = "authorization";
+
+# The prefix used to denote the Basic authentication scheme.
+public const string AUTH_SCHEME_BASIC = "Basic";
+
+# The prefix used to denote the Bearer authentication scheme.
+public const string AUTH_SCHEME_BEARER = "Bearer";
+
+# The permission denied error message.
+public const string PERMISSION_DENIED_ERROR_MSG = "Permission denied";
+
+# The permission denied error message.
+public const string UNAUTHENTICATED_ERROR_MSG = "Unauthenticated";
+
+# Defines the authentication configurations for the HTTP client.
+public type ClientAuthConfig CredentialsConfig|BearerTokenConfig|JwtIssuerConfig|OAuth2GrantConfig;
+
+// Defines the client authentication handlers.
+type ClientAuthHandler ClientBasicAuthHandler|ClientBearerTokenAuthHandler|ClientSelfSignedJwtAuthHandler|ClientOAuth2Handler;
+
+// Logs and prepares the `error` as an `http:ClientAuthError`.
+isolated function prepareClientAuthError(string message, error? err = ()) returns ClientAuthError {
+    log:printError(message, err = err);
+    if (err is error) {
+        return error ClientAuthError(message, err);
+    }
+    return error ClientAuthError(message);
+}
+
+// Extract the credential from `map<string[]>`
+isolated function extractCredential(map<string[]> headers) returns string {
+    string[] authHeader = headers.get(AUTH_HEADER);
+    if (authHeader.length() > 0) {
+        return stringutils:split(<string>authHeader[0], " ")[1];
+    } else {
+        return "";
+    }
+
+}
```

Comment:
If we can update this as follows,  and check for the `()` value at auth_handler and return `UnauthenticatedError` from handler, will avoid the unnecessary call to the auth_provider with empty credentials. I think, HTTP module implementation also should be updated like that. WDYT?
```suggestion
isolated function extractCredential(map<string[]> headers) returns string? {
    string[] authHeader = headers.get(AUTH_HEADER);
    if (authHeader.length() > 0) {
        return stringutils:split(<string>authHeader[0], " ")[1];
    }
}
```

Reply:
Yes, +1. Fixed in here https://github.com/ballerina-platform/module-ballerina-grpc/pull/74/commits/c88f5020d80de8bb530db545064caca0b8cd9317. Slightly change the suggested logic to be compatible with `regex` APIs.